### PR TITLE
Add backend class from qibo

### PR DESCRIPTION
**Pull request by stavros11**
Following qiboteam/qibo#500 this adds the `JITCustomBackend` that was previously in qibo. Since this class inherits the `AbstractBackend` from qibo, qibo should now added in the requirements list for qibojit. 

Note that there is a version 0.1.7 requirement (not yet released), as the new `AbstractCustomOperators` class which is used here is not implemented in earlier versions, it is implemented only in qiboteam/qibo#500. @scarrazza if you believe that this abstraction is a bit of "overkill", in principle we could merge it with the existing `AbstractBackend`. This will relax the version constraint from the qibojit side, but the latest qibo (once released) will still require latest qibojit to work properly.

We may also want to consider expanding the qibojit tests, or moving some backend tests from qibo, since now that the backend class is here.

### CODE REVIEW

**Comment by stavros11** [comment]
@mlazzarin, I moved the regressions for the tests that involve random numbers in the backend classes. Now you can hard-code the results for the rocm case [on this line](https://github.com/qiboteam/qibojit/blob/38845ec7db8f3adbff3e81083fe27da308bc3ff1/src/qibojit/custom_operators/backends.py#L168), see for example the CUDA results a few lines below. Tests should work if you checkout the branches I write [here](https://github.com/qiboteam/qibo/pull/500#issuecomment-955696765). Also, let me know if you agree with this way for treating the regressions.

**Comment by stavros11** [comment]
I added some new tests to increase the coverage to 100%. Once we merge #36 here, this should be good for a qibojit pre-release and check if everything is working with the qibo CI.

**Review by scarrazza — APPROVED** [review]
@stavros11 this too, looks good to go.

File: .github/workflows/rules.yml [hunk 1]
```diff
@@ -8,7 +8,7 @@ jobs:
     strategy:
       matrix:
         os: [ubuntu-latest, macos-latest, windows-latest]
-        python-version: [3.6, 3.7, 3.8, 3.9]
+        python-version: [3.7, 3.8, 3.9]
```
Comment (scarrazza) [inline .github/workflows/rules.yml:11]: @stavros11 @mlazzarin if I remember correctly, the problem with 3.6 is due to the matplotlib dependency.
I don't think qibo should depend on matplotlib, we can include a special installation target ['examples'] which install matplotlib, what do you think?

Comment (stavros11) [reply .github/workflows/rules.yml]: I think the py36 issue is solved if we install qibo from source but using `pip install .`. I am not sure what was causing this issue, the error was indeed related to matplotlib, however it is not appearing in other PRs for example in qibo repository.

Comment (scarrazza) [reply .github/workflows/rules.yml]: Thanks for this, very good that it works.

**Pull request merged**